SCC-2934 - Dynamic schemaName

nonword opened the pull request:
The main functional thing this does is:
 - When doing Avro-decoding on the records in the incoming firehose event, the app no longer uses `process.env.SCHEMA_NAME` to determine what Avro schema to use. It now determines the schema from the `deliverySteamArn` in the event (which contains the kinesis stream name, which conventionally indicates the Avro schema name)

Necessary secondary changes introduced:
 - Update to Node14 (and many related dependency updates)
 - Adds test coverage over sample events
 - Include sample `sam` template for local invocation
 - Add explicit git workflow

Not strictly necessary (but beneficial) changes introduced:
 - Retired use of ESM import/export style because there doesn't appear to be native support in the Node14 Lambda runtime and transpilation using `babel` feels heavy handed just for ESm import/exports (and complicates deployment)
 - Relatedly, removed `babel`
 - Replaced `node-lambda` as a deployment process with `terraform`
 - Added `qa` and `production` config into source control
 - Removed several inessential `scripts` in `package.json`
 - Reworked and simplified `README.md`
 - Added sample events
 - Fixed standardjs linting issues
 - Move `master` to `main`

https://jira.nypl.org/browse/SCC-2934

nonword committed:
Change: dynamically determine schema name

Primarily this changes the handler to determine the relevant schema name
by inspecting the stream on the incoming firehose event.

Secondarily it:
 - Adds c8 for code coverage
 - Removes several deprecated scripts
 - Removes babel
 - Simplifies README
 - Adds sample events
 - Adds qa and production config
 - Changes Node to v14 (from 6)
 - Lints to standardjs
 - Improves test coverage

nonword committed:
Translate to commonjs, add terraform, fix tests

nonword committed:
Remove deployments from package.json

nonword committed:
Fix dependency vulnerabilities

nonword committed:
Change git workflow, fix fixture issue

danamansana commented on .travis.yml:
```diff
@@ -1,5 +1,34 @@
-cache:
-  directories:
-    - "node_modules"
-after_success: npm run coverage
+install:
+- npm i
+before_install:
+- wget https://releases.hashicorp.com/terraform/"$TF_VERSION"/terraform_"$TF_VERSION"_linux_amd64.zip
+- unzip terraform_"$TF_VERSION"_linux_amd64.zip
+- sudo mv terraform /usr/local/bin/
+- rm terraform_"$TF_VERSION"_linux_amd64.zip
+jobs:
+  include:
+  - stage: test
+    script: npm test
+  - stage: deploy qa
+    if: type IN (push) and branch = qa
+    env:
+    - AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID_QA
+    - AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY_QA
+    script:
+    - terraform -chdir=provisioning/qa init -input=false
+    - echo "Deploying to qa"
+    - terraform -chdir=provisioning/qa apply -auto-approve -input=false
+  - stage: deploy production
+    if: type IN (push) and branch = production
+    env:
+    - AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID_PRODUCTION
+    - AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY_PRODUCTION
+    script:
+    - terraform -chdir=provisioning/production init -input=false
+    - echo "Deploying to production"
+    - terraform -chdir=provisioning/production apply -auto-approve -input=false
+env:
+  global:
+  - TF_VERSION=1.0.11
+  - secure: esNcR5939Fd56jFp32PEJ53fKj9Vt0Ehfaru5/qnQrUTTt2Da2ObJ8dXCRa80OPXWURYU4Y8KBOlRvF+Tu1zbE6wrcxGFV9roy+ghqRVRQ/32ZJ28lMwagGu713Cw3LjZlY1JqPsoX7SlXzo1QS9uXlbXiH7kw+VhKfe1/PfiBKCIrkfn6SAs6IM8XdtYw7Q+EsphkALrhkZntAyBbCPkUSLfFNy6Urv2wdCAwFZV3Yemz93g1+4C7j5qtWCTEN0aZen17oLY95IePYlusmpjLiZ67Df8LuwW7WDrA5JppH8zSTM6Hc7vaq8zxVwSCmhYk8Jpbh3oECjHCEtmUvGxMtbps833MpummwWEb5CdcK4ucRSISL/ES+N27dqCafvMOS ...
```
Awesome! Glad to see us moving to Terraform.

danamansana commented on README.md:
```diff
@@ -1,204 +1,68 @@
 # Kinesis Firehose Avro to Json Transformer Lambda
```
Thanks for cleaning this up

danamansana commented on index.js:
```diff
@@ -49,24 +49,43 @@ exports.configHandler = (records, opts, context, callback) => {
 }
 
-exports.handler = (event, context, callback) => {
+/**
```
I'm still a little worried about our reliance on convention here. Should we at least add some error handling/logging in case the schema isn't fetched as expected?

nonword replied:
Good call. I reviewed what will happen if the firehose event stream name implies a nonexistent schema. I added a little more detail to the existing error log and added a specific test over the condition. Essentially, it errors hard and logs out
> `{"level":"ERROR", ,"message":"An error occurred requesting the schema from the NYPL API (https://platform.nypl.org/api/v0.1/current-schemas/NonExistentSchema); the service responded with status code: (404) and status text: (Not Found)",...}`

Additionally:
    This work uncovered that several thrown errors were sending the error
    _message_ as the error callback, which is _fine_ but not quite correct;
    Changed them to pass the whole Error instance back. This required
    updating several tests of error conditions.

danamansana reviewed (commented):
Looks like you did a lot of additional maintenance on this repo with the README, travis, events, etc, so thanks for that. Just one minor comment

nonword committed:
Improve logging around schema lookup failure

nonword committed:
Add test coverage of missing schema, fix error cb

Adds a specific test for handling firehose events tagged with a stream
name for which we don't have an avro schema. Asserts that the invocation
calls back with an error and logs helpful info.

This work uncovered that several thrown errors were sending the error
message as the error callback, which is ~fine~ but not quite correct;
Changed them to pass the whole error back. This required updating
several tests of error conditions.
